src/web/app/pages-admin/admin-search-page/admin-search-page.component.spec.ts

```diff
@@ -555,4 +618,179 @@ describe('AdminSearchPageComponent', () => {
     expect(spyStatusMessageService).toBeCalled();
   });
 
+  it('should generate email when course join email button clicked', () => {
+    const studentResult: StudentAccountSearchResult = {
+      name: 'jack',
+      email: 'email@test.com',
+      googleId: 'googleId',
+      courseId: 'courseId',
+      courseName: 'courseName',
+      institute: 'institute',
+      courseJoinLink: 'courseJoinLink',
+      homePageLink: 'homePageLink',
+      manageAccountLink: 'manageAccountLink',
+      showLinks: true,
+      section: 'section',
+      team: 'team',
+      comments: 'comments',
+      recordsPageLink: 'recordsPageLink',
+      openSessions: DEFAULT_FEEDBACK_SESSION_GROUP,
+      notOpenSessions: DEFAULT_FEEDBACK_SESSION_GROUP,
+      publishedSessions: DEFAULT_FEEDBACK_SESSION_GROUP,
+    };
+    component.students = [studentResult];
+    fixture.detectChanges();
+
+    const spyEmailGenerationService: any = spyOn(emailGenerationService, 'getCourseJoinEmail').and.callFake(
+        (): Observable<Email> => of({
+          recipient: 'Jacky Chan',
+          subject: 'Course join email',
+          content: 'Course join email content',
+        }),
+    );
+
+    const sendButton: any = fixture.debugElement.nativeElement.querySelector('#send-course-join-button');
+    sendButton.click();
+
+    expect(spyEmailGenerationService).toBeCalled();
```

Comment:
Technically the email object is not returned or stored in the component afterwards. So I'm wondering if I am spying on the email generation service, is it possible to retrieve the email's recipient/subject/content?

Reply:
True, this indeed relies `window.location.href` which is not very straightforward to test...

Any suggestions? @ChooJeremy @t-cheepeng @madanalogy

Reply:
@jianhandev You could try to inject a `window` object into the test suite using a `InjectionToken` and test against that.

Some references:
https://angular.io/api/core/InjectionToken
https://jasminexie.github.io/injecting-window-in-an-angular-application/

Reply:
Thanks for the suggestion! Will seriously look into that 👍